File: android/src/main/kotlin/com/example/simple_widget/MigrationHelper.kt
package com.example.simple_widget

import android.content.Context
import android.util.Base64
import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest

object MigrationHelper {

  private const val MIGRATION_PREF = "SimpleWidgetMigration"
  private const val KEY_MIGRATED = "migrated_to_files"

  fun isMigrated(context: Context): Boolean {
    val prefs = context.getSharedPreferences(MIGRATION_PREF, Context.MODE_PRIVATE)
    return prefs.getBoolean(KEY_MIGRATED, false)
  }

  private fun setMigrated(context: Context) {
    val prefs = context.getSharedPreferences(MIGRATION_PREF, Context.MODE_PRIVATE)
    prefs.edit().putBoolean(KEY_MIGRATED, true).apply()
  }

  fun migrateBase64ToFiles(context: Context): Boolean {
    if (isMigrated(context)) {
      return false
    }

    val jsonString = AppSharedPreferences.getTimelinesData(context) ?: run {
      setMigrated(context)
      return false
    }

    try {
      val timelines = JSONArray(jsonString)
      var changed = false
      // Track already-saved hashes to deduplicate shared backgrounds
      val hashToPath = mutableMapOf<String, String>()

      for (i in 0 until timelines.length()) {
        val timeline = timelines.getJSONObject(i)
        val data = timeline.getJSONArray("data")

        for (j in 0 until data.length()) {
          val entry = data.getJSONObject(j)

          val bgValue = entry.optString("background", "")
          if (bgValue.isNotEmpty() && !bgValue.startsWith("widget_images/")) {
            val path = saveBase64AsFile(context, bgValue, hashToPath)
            if (path != null) {
              entry.put("background", path)
              changed = true
            }
          }

          val fgValue = entry.optString("foreground", "")
          if (fgValue.isNotEmpty() && !fgValue.startsWith("widget_images/")) {
            val path = saveBase64AsFile(context, fgValue, hashToPath)
            if (path != null) {
              entry.put("foreground", path)
              changed = true
            }
          }
        }
      }

      if (changed) {
        AppSharedPreferences.save(timelines.toString(), context)
      }

      setMigrated(context)
      return changed
    } catch (e: Exception) {
      e.printStackTrace()
      return false
    }
  }

  private fun saveBase64AsFile(
    context: Context,
    base64String: String,
    hashToPath: MutableMap<String, String>
  ): String? {
    return try {
      val bytes = Base64.decode(base64String, Base64.DEFAULT)
      val hash = md5(bytes)

      // Deduplicate: if we already saved an identical image, reuse its path
      hashToPath[hash]?.let { return it }

      val path = ImageFileManager.saveImage(context, bytes, null)
      hashToPath[hash] = path
      path
    } catch (e: Exception) {
      e.printStackTrace()
      null
    }
  }

  private fun md5(bytes: ByteArray): String {
    val digest = MessageDigest.getInstance("MD5")
    val hashBytes = digest.digest(bytes)
    return hashBytes.joinToString("") { "%02x".format(it) }
  }
}

File: ios/Classes/MigrationHelper.swift
//
//  MigrationHelper.swift
//  simple_widget
//

import Foundation
import CommonCrypto

class MigrationHelper {

  static let shared = MigrationHelper()

  private let migrationKey = "migrated_to_files"

  func isMigrated() -> Bool {
    return UserDefaults.standard.bool(forKey: migrationKey)
  }

  private func setMigrated() {
    UserDefaults.standard.set(true, forKey: migrationKey)
  }

  func migrateBase64ToFiles() -> Bool {
    if isMigrated() {
      return false
    }

    guard let jsonString = Storage.shared.getTimelinesData(),
          let jsonData = jsonString.data(using: .utf8) else {
      setMigrated()
      return false
    }

    do {
      guard var timelines = try JSONSerialization.jsonObject(with: jsonData) as? [[String: Any]] else {
        setMigrated()
        return false
      }

      var changed = false
      var hashToPath: [String: String] = [:]

      for i in 0..<timelines.count {
        guard var dataArray = timelines[i]["data"] as? [[String: Any]] else { continue }

        for j in 0..<dataArray.count {
          if let bgValue = dataArray[j]["background"] as? String,
             !bgValue.isEmpty,
             !bgValue.hasPrefix("widget_images/") {
            if let path = saveBase64AsFile(bgValue, hashToPath: &hashToPath) {
              dataArray[j]["background"] = path
              changed = true
            }
          }

          if let fgValue = dataArray[j]["foreground"] as? String,
             !fgValue.isEmpty,
             !fgValue.hasPrefix("widget_images/") {
            if let path = saveBase64AsFile(fgValue, hashToPath: &hashToPath) {
              dataArray[j]["foreground"] = path
              changed = true
            }
          }
        }

        timelines[i]["data"] = dataArray
      }

      if changed {
        let newData = try JSONSerialization.data(withJSONObject: timelines)
        if let newString = String(data: newData, encoding: .utf8) {
          Storage.shared.save(newString)
        }
      }

      setMigrated()
      return changed
    } catch {
      print("Migration failed: \(error)")
      return false
    }
  }

  private func saveBase64AsFile(_ base64String: String, hashToPath: inout [String: String]) -> String? {
    guard let bytes = Data(base64Encoded: base64String) else { return nil }

    let hash = md5(bytes)
    if let existingPath = hashToPath[hash] {
      return existingPath
    }

    guard let path = ImageFileManager.shared.saveImage(bytes, filename: nil) else { return nil }
    hashToPath[hash] = path
    return path
  }

  private func md5(_ data: Data) -> String {
    var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
    data.withUnsafeBytes { body in
      _ = CC_MD5(body.baseAddress, CC_LONG(data.count), &digest)
    }
    return digest.map { String(format: "%02x", $0) }.joined()
  }
}
